src/implementations/astroport/astroport.rs

```diff
@@ -0,0 +1,121 @@
+use cosmwasm_std::{to_binary, Addr, CosmosMsg, Decimal, Empty, Response, StdResult, WasmMsg};
+use cosmwasm_std::{Deps, Event, Uint128};
+use cw_asset::{Asset, AssetInfo, AssetInfoBase, AssetList};
+
+use astroport_core::asset::{Asset as AstroAsset, AssetInfo as AstroAssetInfo};
+use astroport_core::pair::ExecuteMsg as PairExecMsg;
+
+use crate::pool::Pool;
+use crate::CwDexError;
+
+pub struct AstroportPool {
+    /// Information about assets in the pool
+    pub asset_infos: Vec<AssetInfo>,
+    /// The token contract code ID used for the tokens in the pool
+    pub token_code_id: u64,
+    /// The factory contract address
+    pub factory_addr: String,
+}
+
+impl Pool for AstroportPool {
+    fn provide_liquidity(
+        &self,
+        _deps: Deps,
+        assets: AssetList,
+        recipient: Addr,
+        slippage_tolerance: Option<Decimal>,
+    ) -> Result<Response, CwDexError> {
+        let mut astro_assets: Vec<AstroAsset> = vec![];
+        for asset_base in assets.into_iter() {
+            let info = match &asset_base.info {
+                AssetInfoBase::Native(denom) => Ok(AstroAssetInfo::NativeToken {
+                    denom: denom.to_string(),
+                }),
+                AssetInfoBase::Cw20(addr) => Ok(AstroAssetInfo::Token {
+                    contract_addr: addr.to_owned(),
+                }),
+                AssetInfoBase::Cw1155(addr, _) => Ok(AstroAssetInfo::Token {
+                    contract_addr: addr.to_owned(),
+                }),
+                x => Err(CwDexError::InvalidInAsset {
+                    a: asset_base.to_owned(),
+                }),
+            }?;
+            let amount = asset_base.amount;
+            astro_assets.push(AstroAsset {
+                info,
+                amount,
+            })
+        }
+        let msg = PairExecMsg::ProvideLiquidity {
+            assets: astro_assets.clone(),
+            slippage_tolerance,
+            auto_stake: Some(false), // Should this be true?
+            receiver: Some(recipient.to_string()),
+        };
+        let liquidity = CosmosMsg::<Empty>::Wasm(WasmMsg::Execute {
+            contract_addr: self.factory_addr.to_string(),
```

Comment:
we should call provide_liquidity on the pair. So we need a `pair_addr` in the `AstroportPool` struct

Reply:
ACK